core/src/main/java/org/apache/iceberg/schema/UnionByNameVisitor.java

```diff
@@ -0,0 +1,206 @@
+          }
+        });
+
+    return false;
+  }
+
+  @Override
+  public Boolean field(Types.NestedField field, Integer partnerId, Boolean isFieldMissing) {
+    return partnerId == null;
+  }
+
+  @Override
+  public Boolean list(Types.ListType list, Integer partnerId, Boolean isElementMissing) {
+    if (partnerId == null) {
+      return true;
+    }
+
+    Preconditions.checkState(!isElementMissing, "Error traversing schemas: element is missing, but list is present");
+
+    Types.ListType partnerList = findFieldType(partnerId).asListType();
+    updateColumn(list.fields().get(0), partnerList.fields().get(0));
+
+    return false;
+  }
+
+  @Override
+  public Boolean map(Types.MapType map, Integer partnerId, Boolean isKeyMissing, Boolean isValueMissing) {
+    if (partnerId == null) {
+      return true;
+    }
+
+    Preconditions.checkState(!isKeyMissing, "Error traversing schemas: key is missing, but map is present");
+    Preconditions.checkState(!isValueMissing, "Error traversing schemas: value is missing, but map is present");
+
+    Types.MapType partnerMap = findFieldType(partnerId).asMapType();
+    updateColumn(map.fields().get(0), partnerMap.fields().get(0));
+    updateColumn(map.fields().get(1), partnerMap.fields().get(1));
+
+    return false;
+  }
+
+  @Override
+  public Boolean primitive(Type.PrimitiveType primitive, Integer partnerId) {
+    return partnerId == null;
+  }
+
+  private Type findFieldType(int fieldId) {
+    if (fieldId == -1) {
+      return partnerSchema.asStruct();
+    } else {
+      return partnerSchema.findField(fieldId).type();
+    }
+  }
+
+  private void addColumn(int parentId, Types.NestedField field) {
+    String parentName = partnerSchema.findColumnName(parentId);
+    api.addColumn(parentName, field.name(), field.type(), field.doc());
+  }
+
+  private void updateColumn(Types.NestedField field, Types.NestedField existingField) {
+    String fullName = partnerSchema.findColumnName(existingField.fieldId());
+
+    boolean needsOptionalUpdate = field.isOptional() && existingField.isRequired();
+    boolean needsTypeUpdate = field.type().isPrimitiveType() && !field.type().equals(existingField.type());
+    boolean needsDocUpdate = field.doc() != null && !field.doc().equals(existingField.doc());
+
+    if (needsOptionalUpdate) {
+      api.makeColumnOptional(fullName);
+    }
+
+    if (needsTypeUpdate) {
+      api.updateColumn(fullName, field.type().asPrimitiveType());
+    }
+
+    if (needsDocUpdate) {
+      api.updateColumnDoc(fullName, field.doc());
+    }
+  }
+
+  private static class PartnerIdByNameAccessors implements PartnerAccessors<Integer> {
+    private final Schema partnerSchema;
+
+    private PartnerIdByNameAccessors(Schema partnerSchema) {
+      this.partnerSchema = partnerSchema;
+    }
+
+    @Override
+    public Integer fieldPartner(Integer partnerFieldId, int fieldId, String name) {
+      Types.StructType struct;
+      if (partnerFieldId == -1) {
```

Comment:
@rdblue I think this is pretty much the only change I've made to your poc code - tests are passing just fine - will add some more from past use-cases - I realise that an exhaustive test suite is pretty much impossible for this feature but I'll try to cover both basic and complex use-cases

Reply:
I think this is correct.